staging/src/k8s.io/client-go/tools/cache/shared_informer_health.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+func newDefaultListWatchHealth(stopCh <-chan struct{}, errorReceiverCh <-chan int) *listWatchHealth {
+	return &listWatchHealth{
+		minTimeToFail:       time.Duration(5 * time.Second),
+		minConsistentErrors: 10,
+		consistentErrorsCh:  errorReceiverCh,
+		stopCh:              stopCh,
+	}
+}
+
+type listWatchHealth struct {
+	lock         sync.Mutex
+	startingTime time.Time
+
+	// minTimeToFail minimum time duration after which we start to fail
+	minTimeToFail time.Duration
+	// minConsistentError minimum number of errors before we start to fail
+	minConsistentErrors int
+	// consistentErrors holds the number of errors seen so far
+	consistentErrors int
+
+	// consistentErrorsCh channel on which we receive List&Watch related errors
+	// it is meant to represent a gauge, that is a number that can go up and down
+	consistentErrorsCh <-chan int
```

Comment:
the health check would be more useful if you also tracked the last 5 errors or so and produced them on failing health so that callers could see why on detailed health.